File: flutter_app/android/app/src/main/kotlin/com/drn/app/ProxyBridge.kt
package com.drn.app

import android.util.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class ProxyBridge {
    private val TAG = "Dr.N-Bridge"

    fun register(flutterEngine: FlutterEngine) {
        val channel = MethodChannel(flutterEngine.dartExecutor.binaryMessenger, "drn.app/proxy")
        Log.d(TAG, "Registering MethodChannel: drn.app/proxy")

        channel.setMethodCallHandler { call: MethodCall, result: MethodChannel.Result ->
            Log.d(TAG, "Method called: ${call.method}")
            try {
                when (call.method) {
                    "getServers" -> {
                        result.success("[]")
                    }
                    "getStatus" -> {
                        result.success("""{"running":false,"note":"Go backend not loaded"}""")
                    }
                    "addServer" -> {
                        result.success(null)
                    }
                    "removeServer" -> {
                        result.success(null)
                    }
                    "startProxy" -> {
                        val address = call.argument<String>("address") ?: ""
                        val port = call.argument<String>("port") ?: "19132"
                        Log.d(TAG, "startProxy called: $address:$port")

                        // Try to use Go backend if available, otherwise return error
                        try {
                            val core = drnbind.Core()
                            core.startProxy(address, port)
                            Log.d(TAG, "Proxy started via Go backend")
                            result.success(null)
                        } catch (e: Throwable) {
                            Log.e(TAG, "Go backend failed: ${e.message}")
                            result.error(
                                "BACKEND_UNAVAILABLE",
                                "Go proxy backend not available: ${e.message}",
                                null
                            )
                        }
                    }
                    "stopProxy" -> {
                        try {
                            val core = drnbind.Core()
                            core.stopProxy()
                            result.success(null)
                        } catch (e: Throwable) {
                            result.success(null) // Not running is fine
                        }
                    }
                    else -> result.notImplemented()
                }
            } catch (e: Exception) {
                Log.e(TAG, "Unhandled error: ${e.message}")
                result.error("ERROR", e.message, null)
            }
        }
        Log.d(TAG, "MethodChannel registered successfully")
    }
}

File: flutter_app/ios/Runner/ProxyBridge.swift
import Flutter
import UIKit
import drnbind

public class SwiftProxyBridge {
    private let core = Core()

    func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "drn.app/proxy",
            binaryMessenger: registrar.messenger()
        )

        channel.setMethodCallHandler { [weak self] call, result in
            guard let self = self else { return }

            do {
                switch call.method {
                case "getServers":
                    result(self.core.getServersJSON())

                case "getStatus":
                    result(self.core.getStatusJSON())

                case "addServer":
                    let args = call.arguments as? [String: String] ?? [:]
                    self.core.addServer(
                        args["name"] ?? "",
                        address: args["address"] ?? "",
                        port: args["port"] ?? "19132"
                    )
                    result(nil)

                case "removeServer":
                    let args = call.arguments as? [String: String] ?? [:]
                    self.core.removeServer(args["id"] ?? "")
                    result(nil)

                case "startProxy":
                    let args = call.arguments as? [String: String] ?? [:]
                    try self.core.startProxy(
                        args["address"] ?? "",
                        serverPort: args["port"] ?? "19132"
                    )
                    result(nil)

                case "stopProxy":
                    try self.core.stopProxy()
                    result(nil)

                default:
                    result(FlutterMethodNotImplemented)
                }
            } catch {
                result(FlutterError(
                    code: "PROXY_ERROR",
                    message: error.localizedDescription,
                    details: nil
                ))
            }
        }
    }
}

// Plugin registration for iOS
public class DrNPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let bridge = SwiftProxyBridge()
        bridge.register(with: registrar)
    }
}